### PR TITLE
Improve documentation for built-in macros

### PR DESCRIPTION
**Pull request by petrochenkov**
This is the `libcore` part of https://github.com/rust-lang/rust/pull/62086.
Right now the only effect is improved documentation.

The changes in the last few commits are required to make the `libcore` change compile successfully.

### CODE REVIEW

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -4008,7 +4008,7 @@ impl<'a> LoweringContext<'a> {
         let attrs = self.lower_attrs(&i.attrs);
         if let ItemKind::MacroDef(ref def) = i.node {
             if !def.legacy || attr::contains_name(&i.attrs, sym::macro_export) ||
-                              attr::contains_name(&i.attrs, sym::rustc_doc_only_macro) {
+                              attr::contains_name(&i.attrs, sym::rustc_builtin_macro) {
```
Comment (eddyb) [inline src/librustc/hir/lowering.rs]: Is this needed anymore? The `macro_rules!` -> `macro` change would mean `!def.legacy` is now `true`.

File: src/librustc_resolve/macros.rs [hunk 2]
```diff
@@ -323,21 +323,13 @@ impl<'a> Resolver<'a> {
                     let features = self.session.features_untracked();
                     if attr_kind == NonMacroAttrKind::Custom {
                         assert!(path.segments.len() == 1);
-                        let name = path.segments[0].ident.as_str();
-                        if name.starts_with("rustc_") {
-                            if !features.rustc_attrs {
-                                let msg = "unless otherwise specified, attributes with the prefix \
-                                           `rustc_` are reserved for internal compiler diagnostics";
```
Comment (eddyb) [inline src/librustc_resolve/macros.rs]: Where did this go?

File: src/libstd/macros.rs [hunk 3]
```diff
@@ -410,7 +410,7 @@ mod builtin {
     ///
     /// [`panic!`]: ../std/macro.panic.html
     #[stable(feature = "compile_error_macro", since = "1.20.0")]
-    #[rustc_doc_only_macro]
+    #[rustc_builtin_macro]
     macro_rules! compile_error {
```
Comment (eddyb) [inline src/libstd/macros.rs]: Oh, these are unchanged. What needs to happen before these can just be reexports of the `libcore` ones?

File: src/libsyntax/feature_gate.rs [hunk 4]
```diff
@@ -1351,6 +1365,26 @@ pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
...
+
+    (sym::rustc_promotable, Whitelisted, template!(Word), Gated(Stability::Unstable,
+                                                sym::rustc_attrs,
+                                                "internal implementation detail",
+                                                cfg_fn!(rustc_attrs))),
+
+    (sym::rustc_allow_const_fn_ptr, Whitelisted, template!(Word), Gated(Stability::Unstable,
+                                                sym::rustc_attrs,
+                                                "internal implementation detail",
+                                                cfg_fn!(rustc_attrs))),
```
Comment (eddyb) [inline src/libsyntax/feature_gate.rs]: Oh, you whitelisted all the existing `rustc_*` attributes?

File: src/librustc/hir/lowering.rs (hunk 1 above)
Comment (petrochenkov) [reply src/librustc/hir/lowering.rs]: No, this is no longer needed.

File: src/libsyntax/feature_gate.rs (hunk 4 above)
Comment (petrochenkov) [reply src/libsyntax/feature_gate.rs]: Yep.
If an attribute is not defined/registered in any way, then it goes through the "`feature(custom_attribute)` fallback" (https://github.com/rust-lang/rust/issues/29642), something I want to remove from resolve (https://github.com/rust-lang/rust/pull/61660, https://github.com/rust-lang/rust/pull/57921).

Also, it's just useful for documentation.
https://github.com/rust-lang/rust/pull/62133 does the same thing.

File: src/libstd/macros.rs (hunk 3 above)
Comment (petrochenkov) [reply src/libstd/macros.rs]: `rustc_builtin_macro` stubs acting like the real macros.

https://github.com/rust-lang/rust/pull/62086 removes the duplicate macros from libstd, but also `pub use`s their libcore versions from libstd's root (for documentation). 
...And stub macros `pub use`d in libstd root break all the dependent crates that inject `#[macro_use] extern crate std` during rustdoc build.

Alternatively, some `doc` attribute for selectively documenting private items would also solve the problem (`pub use` could be replaced with `#[doc(force)] use` then).

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/62442#issuecomment-508925513, @bors r-

**Comment by petrochenkov** [comment]
>Failed in #62442 (comment)

Ugh, unstable diagnostic output.

**Comment by petrochenkov** [comment]
>unstable diagnostic output.

Order in which `mir_const_qualif` visits items somehow depends on macro loading order.
I didn't investigate that in detail, but I changed the last commit to not load macros in `check_reserved_macro_name` unless absolutely necessary instead, thus restoring the old macro loading order.

@bors r=eddyb

**Pull request merged**